File: ios/Classes/EzoicFlutterSdkPlugin.swift
import Flutter
import UIKit
import EzoicAdsSDKBinary

public class EzoicFlutterSdkPlugin: NSObject, FlutterPlugin {
  private var messenger: FlutterBinaryMessenger?

  /// Loaded rewarded ads awaiting `show`, keyed by ad unit id.
  private var rewardedAds: [Int: EzoicRewardedAd] = [:]

  /// Per-ad event channels used to surface lifecycle callbacks to Dart.
  private var rewardedChannels: [Int: FlutterMethodChannel] = [:]

  /// In-flight `show` calls, keyed by ad unit id.
  private var pendingShows: [Int: PendingRewardShow] = [:]

  private final class PendingRewardShow {
    let result: FlutterResult
    var reward: EzoicReward?
    var settled = false
    init(_ result: @escaping FlutterResult) { self.result = result }
  }

  /// Loaded interstitial ads awaiting `show`, keyed by ad unit id.
  private var interstitialAds: [Int: EzoicInterstitialAd] = [:]

  /// Per-ad event channels used to surface interstitial callbacks to Dart.
  private var interstitialChannels: [Int: FlutterMethodChannel] = [:]

  /// In-flight interstitial `show` calls, keyed by ad unit id.
  private var pendingInterstitialShows: [Int: PendingInterstitialShow] = [:]

  private final class PendingInterstitialShow {
    let result: FlutterResult
    var settled = false
    init(_ result: @escaping FlutterResult) { self.result = result }
  }

  /// Ad unit ids with an in-flight rewarded `load`.
  private var loadingRewarded: Set<Int> = []

  /// Ad unit ids with an in-flight interstitial `load`.
  private var loadingInterstitial: Set<Int> = []

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "com.ezoic/ezoic_flutter_sdk", binaryMessenger: registrar.messenger())
    let instance = EzoicFlutterSdkPlugin()
    instance.messenger = registrar.messenger()
    registrar.addMethodCallDelegate(instance, channel: channel)

    let factory = EzoicBannerViewFactory(messenger: registrar.messenger())
    registrar.register(factory, withId: "com.ezoic/ezoic_banner_view")
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "initialize":
      guard let args = call.arguments as? [String: Any],
            let domain = args["domain"] as? String, !domain.isEmpty else {
        result(FlutterError(code: "EzoicAds", message: "initialize requires a non-empty `domain`.", details: nil))
        return
      }
      let config = EzoicConfiguration(
        domain: domain,
        autoReadConsent: args["autoReadConsent"] as? Bool ?? true,
        subjectToCOPPA: args["subjectToCOPPA"] as? Bool ?? false,
        requestATTBeforeAds: args["requestATTBeforeAds"] as? Bool ?? true,
        debugEnabled: args["debugEnabled"] as? Bool ?? false,
        testMode: args["testMode"] as? Bool ?? false
      )
      EzoicAds.shared.initialize(with: config) { r in
        switch r {
        case .success: result(nil)
        case .failure(let e): result(FlutterError(code: "EzoicAds", message: e.localizedDescription, details: e.code))
        }
      }
    case "setGDPRConsent":
      let args = call.arguments as? [String: Any]
      EzoicAds.shared.setGDPRConsent(applies: args?["applies"] as? Bool ?? false,
                                     consentString: args?["consentString"] as? String)
      result(nil)
    case "setGPPConsent":
      let args = call.arguments as? [String: Any]
      EzoicAds.shared.setGPPConsent(gppString: args?["gppString"] as? String,
                                    sectionIds: args?["sectionIds"] as? String)
      result(nil)
    case "setSubjectToCOPPA":
      let args = call.arguments as? [String: Any]
      EzoicAds.shared.setSubjectToCOPPA(args?["value"] as? Bool ?? false)
      result(nil)
    case "trackPageview":
      EzoicAds.shared.trackPageview { success in result(success) }
    case "loadRewardedAd":
      handleLoadRewardedAd(call, result)
    case "showRewardedAd":
      handleShowRewardedAd(call, result)
    case "loadInterstitialAd":
      handleLoadInterstitialAd(call, result)
    case "showInterstitialAd":
      handleShowInterstitialAd(call, result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func handleLoadRewardedAd(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
          let adUnitIdentifier = args["adUnitIdentifier"] as? String,
          let id = Int(adUnitIdentifier) else {
      result(FlutterError(code: "EzoicAds", message: "Invalid adUnitIdentifier.", details: nil))
      return
    }
    if rewardedAds[id] != nil || loadingRewarded.contains(id) {
      result(FlutterError(code: "EzoicAds",
                          message: "An ad is already loaded/loading for ad unit \(adUnitIdentifier)",
                          details: nil))
      return
    }
    loadingRewarded.insert(id)
    EzoicRewardedAd.load(adUnitIdentifier: id) { [weak self] r in
      guard let self = self else { return }
      self.loadingRewarded.remove(id)
      switch r {
      case .success(let ad):
        ad.delegate = self
        self.rewardedAds[id] = ad
        if let messenger = self.messenger, self.rewardedChannels[id] == nil {
          self.rewardedChannels[id] = FlutterMethodChannel(
            name: "com.ezoic/ezoic_rewarded_ad_\(id)", binaryMessenger: messenger)
        }
        result(nil)
      case .failure(let e):
        result(FlutterError(code: "EzoicAds", message: e.localizedDescription, details: e.code))
      }
    }
  }

  private func handleShowRewardedAd(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
          let adUnitIdentifier = args["adUnitIdentifier"] as? String,
          let id = Int(adUnitIdentifier), let ad = rewardedAds[id] else {
      result(FlutterError(code: "EzoicAds", message: "Rewarded ad not loaded.", details: nil))
      return
    }
    if pendingShows[id] != nil {
      result(FlutterError(code: "EzoicAds",
                          message: "A show is already in progress for ad unit \(adUnitIdentifier)",
                          details: nil))
      return
    }
    pendingShows[id] = PendingRewardShow(result)
    // Presenting from nil lets GMA use the application's top view controller.
    ad.show(from: nil) { [weak self] reward in
      self?.pendingShows[id]?.reward = reward
    }
  }

  private func emit(_ id: Int, _ method: String, _ args: Any? = nil) {
    rewardedChannels[id]?.invokeMethod(method, arguments: args)
  }

  private func handleLoadInterstitialAd(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
          let adUnitIdentifier = args["adUnitIdentifier"] as? String,
          let id = Int(adUnitIdentifier) else {
      result(FlutterError(code: "EzoicAds", message: "Invalid adUnitIdentifier.", details: nil))
      return
    }
    if interstitialAds[id] != nil || loadingInterstitial.contains(id) {
      result(FlutterError(code: "EzoicAds",
                          message: "An ad is already loaded/loading for ad unit \(adUnitIdentifier)",
                          details: nil))
      return
    }
    loadingInterstitial.insert(id)
    EzoicInterstitialAd.load(adUnitIdentifier: id) { [weak self] r in
      guard let self = self else { return }
      self.loadingInterstitial.remove(id)
      switch r {
      case .success(let ad):
        ad.delegate = self
        self.interstitialAds[id] = ad
        if let messenger = self.messenger, self.interstitialChannels[id] == nil {
          self.interstitialChannels[id] = FlutterMethodChannel(
            name: "com.ezoic/ezoic_interstitial_ad_\(id)", binaryMessenger: messenger)
        }
        result(nil)
      case .failure(let e):
        result(FlutterError(code: "EzoicAds", message: e.localizedDescription, details: e.code))
      }
    }
  }

  private func handleShowInterstitialAd(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
          let adUnitIdentifier = args["adUnitIdentifier"] as? String,
          let id = Int(adUnitIdentifier), let ad = interstitialAds[id] else {
      result(FlutterError(code: "EzoicAds", message: "Interstitial ad not loaded.", details: nil))
      return
    }
    if pendingInterstitialShows[id] != nil {
      result(FlutterError(code: "EzoicAds",
                          message: "A show is already in progress for ad unit \(adUnitIdentifier)",
                          details: nil))
      return
    }
    pendingInterstitialShows[id] = PendingInterstitialShow(result)
    // Native show(from:) has no completion handler, so the show promise is
    // settled from the delegate (dismiss = resolve, failed-to-present = reject).
    // Presenting from nil lets GMA use the application's top view controller.
    ad.show(from: nil)
  }

  private func emitInterstitial(_ id: Int, _ method: String, _ args: Any? = nil) {
    interstitialChannels[id]?.invokeMethod(method, arguments: args)
  }
}

// MARK: - EzoicRewardedAdDelegate

extension EzoicFlutterSdkPlugin: EzoicRewardedAdDelegate {

  public func rewardedAdDidPresent(_ rewardedAd: EzoicRewardedAd) {
    emit(rewardedAd.adUnitIdentifier, "onShown")
  }

  public func rewardedAd(_ rewardedAd: EzoicRewardedAd, didFailToPresentWithError error: EzoicError) {
    let id = rewardedAd.adUnitIdentifier
    emit(id, "onFailedToShow", ["message": error.localizedDescription, "code": error.code])
    rewardedAds.removeValue(forKey: id)
    if let pending = pendingShows.removeValue(forKey: id), !pending.settled {
      pending.settled = true
      pending.result(FlutterError(code: "EzoicAds", message: error.localizedDescription, details: error.code))
    }
  }

  public func rewardedAdDidRecordImpression(_ rewardedAd: EzoicRewardedAd) {
    emit(rewardedAd.adUnitIdentifier, "onImpression")
  }

  public func rewardedAdDidRecordClick(_ rewardedAd: EzoicRewardedAd) {
    emit(rewardedAd.adUnitIdentifier, "onClicked")
  }

  public func rewardedAd(_ rewardedAd: EzoicRewardedAd, userDidEarn reward: EzoicReward) {
    emit(rewardedAd.adUnitIdentifier, "onUserEarnedReward", ["type": reward.type, "amount": reward.amount])
    pendingShows[rewardedAd.adUnitIdentifier]?.reward = reward
  }

  public func rewardedAdDidDismiss(_ rewardedAd: EzoicRewardedAd) {
    let id = rewardedAd.adUnitIdentifier
    emit(id, "onDismissed")
    let pending = pendingShows.removeValue(forKey: id)
    rewardedAds.removeValue(forKey: id)
    if let pending = pending, !pending.settled {
      pending.settled = true
      let reward = pending.reward
      pending.result([
        "earned": reward != nil,
        "type": reward?.type ?? "",
        "amount": reward?.amount ?? 0
      ])
    }
  }
}

// MARK: - EzoicInterstitialAdDelegate

extension EzoicFlutterSdkPlugin: EzoicInterstitialAdDelegate {

  public func interstitialAdDidPresent(_ interstitialAd: EzoicInterstitialAd) {
    emitInterstitial(interstitialAd.adUnitIdentifier, "onShown")
  }

  public func interstitialAd(_ interstitialAd: EzoicInterstitialAd, didFailToPresentWithError error: EzoicError) {
    let id = interstitialAd.adUnitIdentifier
    emitInterstitial(id, "onFailedToShow", ["message": error.localizedDescription, "code": error.code])
    interstitialAds.removeValue(forKey: id)
    if let pending = pendingInterstitialShows.removeValue(forKey: id), !pending.settled {
      pending.settled = true
      pending.result(FlutterError(code: "EzoicAds", message: error.localizedDescription, details: error.code))
    }
  }

  public func interstitialAdDidRecordImpression(_ interstitialAd: EzoicInterstitialAd) {
    emitInterstitial(interstitialAd.adUnitIdentifier, "onImpression")
  }

  public func interstitialAdDidRecordClick(_ interstitialAd: EzoicInterstitialAd) {
    emitInterstitial(interstitialAd.adUnitIdentifier, "onClicked")
  }

  public func interstitialAdDidDismiss(_ interstitialAd: EzoicInterstitialAd) {
    let id = interstitialAd.adUnitIdentifier
    emitInterstitial(id, "onDismissed")
    let pending = pendingInterstitialShows.removeValue(forKey: id)
    interstitialAds.removeValue(forKey: id)
    if let pending = pending, !pending.settled {
      pending.settled = true
      pending.result(nil)
    }
  }
}

File: android/src/main/kotlin/com/ezoic/ezoic_flutter_sdk/EzoicFlutterSdkPlugin.kt
package com.ezoic.ezoic_flutter_sdk

import android.app.Activity
import android.app.Application
import android.content.Context
import android.os.Handler
import android.os.Looper
import com.ezoic.ads.sdk.adunits.EzoicInstreamAd
import com.ezoic.ads.sdk.adunits.EzoicInstreamAdListener
import com.ezoic.ads.sdk.adunits.EzoicInterstitialAd
import com.ezoic.ads.sdk.adunits.EzoicInterstitialAdListenerAdapter
import com.ezoic.ads.sdk.adunits.EzoicReward
import com.ezoic.ads.sdk.adunits.EzoicRewardedAd
import com.ezoic.ads.sdk.adunits.EzoicRewardedAdListenerAdapter
import com.ezoic.ads.sdk.core.EzoicAds
import com.ezoic.ads.sdk.core.EzoicConfiguration
import com.ezoic.ads.sdk.core.EzoicError
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class EzoicFlutterSdkPlugin : FlutterPlugin, ActivityAware, MethodChannel.MethodCallHandler {
  private lateinit var channel: MethodChannel
  private lateinit var appContext: Context
  private lateinit var messenger: BinaryMessenger
  private val mainHandler = Handler(Looper.getMainLooper())

  /** The host Activity, available while the plugin is attached to one. */
  private var activity: Activity? = null

  /** Loaded rewarded ads awaiting `show`, keyed by ad unit id. */
  private val rewardedAds = HashMap<String, EzoicRewardedAd>()

  /** Per-ad event channels used to surface lifecycle callbacks to Dart. */
  private val rewardedChannels = HashMap<String, MethodChannel>()

  /** In-flight `show` calls, keyed by ad unit id. */
  private val pendingShows = HashMap<String, RewardShow>()

  private class RewardShow(val result: MethodChannel.Result) {
    var settled = false
    var reward: EzoicReward? = null
  }

  /** Loaded interstitial ads awaiting `show`, keyed by ad unit id. */
  private val interstitialAds = HashMap<String, EzoicInterstitialAd>()

  /** Per-ad event channels used to surface interstitial callbacks to Dart. */
  private val interstitialChannels = HashMap<String, MethodChannel>()

  /** In-flight interstitial `show` calls, keyed by ad unit id. */
  private val pendingInterstitialShows = HashMap<String, InterstitialShow>()

  private class InterstitialShow(val result: MethodChannel.Result) {
    var settled = false
  }

  /** Ad unit ids with an in-flight rewarded `load`. */
  private val loadingRewarded = HashSet<String>()

  /** Ad unit ids with an in-flight interstitial `load`. */
  private val loadingInterstitial = HashSet<String>()

  /** Live instream controllers, keyed by ad unit id. Multi-use — reused across loads. */
  private val instreamAds = HashMap<Int, EzoicInstreamAd>()

  /**
   * In-flight instream `load` calls, keyed by ad unit id. The native load
   * silently no-ops while loading, which would hang the Dart future, so an
   * overlapping load is rejected here instead. Also lets `destroyInstreamAd`
   * settle a pending load's result: the native SDKs ignore callbacks for a
   * destroyed ad, so without this the Dart `load()` future would hang forever.
   */
  private val loadingInstream = HashMap<Int, InstreamLoad>()

  private class InstreamLoad(val result: MethodChannel.Result) {
    var settled = false
  }

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    appContext = binding.applicationContext
    messenger = binding.binaryMessenger
    channel = MethodChannel(messenger, "com.ezoic/ezoic_flutter_sdk")
    channel.setMethodCallHandler(this)
    binding.platformViewRegistry.registerViewFactory(
      "com.ezoic/ezoic_banner_view",
      EzoicBannerViewFactory(messenger)
    )
    binding.platformViewRegistry.registerViewFactory(
      "com.ezoic/ezoic_native_ad_view",
      EzoicNativeAdViewFactory(messenger)
    )
    binding.platformViewRegistry.registerViewFactory(
      "com.ezoic/ezoic_outstream_ad_view",
      EzoicOutstreamAdViewFactory(messenger)
    )
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
    rewardedChannels.values.forEach { it.setMethodCallHandler(null) }
    rewardedChannels.clear()
    rewardedAds.clear()
    pendingShows.clear()
    loadingRewarded.clear()
    interstitialChannels.values.forEach { it.setMethodCallHandler(null) }
    interstitialChannels.clear()
    interstitialAds.clear()
    pendingInterstitialShows.clear()
    loadingInterstitial.clear()
    loadingInstream.values.forEach { holder ->
      if (!holder.settled) {
        holder.settled = true
        holder.result.error("EzoicAds", "Engine detached while an instream ad was loading", null)
      }
    }
    instreamAds.values.forEach { it.destroy() }
    instreamAds.clear()
    loadingInstream.clear()
  }

  // ActivityAware — capture the Activity so rewarded ads can be presented.

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    activity = binding.activity
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    activity = binding.activity
  }

  override fun onDetachedFromActivityForConfigChanges() {
    activity = null
  }

  override fun onDetachedFromActivity() {
    activity = null
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    when (call.method) {
      "initialize" -> {
        val domain = call.argument<String>("domain")
        if (domain.isNullOrEmpty()) {
          result.error("EzoicAds", "initialize requires a non-empty `domain`.", null); return
        }
        val app = appContext as? Application
        if (app == null) { result.error("EzoicAds", "No Application context available.", null); return }
        val config = EzoicConfiguration(
          domain = domain,
          autoReadConsent = call.argument<Boolean>("autoReadConsent") ?: true,
          subjectToCOPPA = call.argument<Boolean>("subjectToCOPPA") ?: false,
          requestATTBeforeAds = call.argument<Boolean>("requestATTBeforeAds") ?: true,
          debugEnabled = call.argument<Boolean>("debugEnabled") ?: false,
          testMode = call.argument<Boolean>("testMode") ?: false
        )
        EzoicAds.instance.initialize(app, config) { r ->
          r.onSuccess { result.success(null) }
            .onFailure { e -> result.error("EzoicAds", e.message, e.toString()) }
        }
      }
      "setGDPRConsent" -> {
        EzoicAds.instance.setGDPRConsent(
          call.argument<Boolean>("applies") ?: false, call.argument<String>("consentString"))
        result.success(null)
      }
      "setGPPConsent" -> {
        EzoicAds.instance.setGPPConsent(
          call.argument<String>("gppString"), call.argument<String>("sectionIds"))
        result.success(null)
      }
      "setSubjectToCOPPA" -> {
        EzoicAds.instance.setSubjectToCOPPA(call.argument<Boolean>("value") ?: false)
        result.success(null)
      }
      "trackPageview" -> EzoicAds.instance.trackPageview { success -> result.success(success) }
      "loadRewardedAd" -> handleLoadRewardedAd(call, result)
      "showRewardedAd" -> handleShowRewardedAd(call, result)
      "loadInterstitialAd" -> handleLoadInterstitialAd(call, result)
      "showInterstitialAd" -> handleShowInterstitialAd(call, result)
      "loadInstreamAd" -> handleLoadInstreamAd(call, result)
      "getInstreamNextAdTagUrl" -> handleGetInstreamNextAdTagUrl(call, result)
      "reportInstreamImpression" -> handleReportInstreamImpression(call, result)
      "destroyInstreamAd" -> handleDestroyInstreamAd(call, result)
      else -> result.notImplemented()
    }
  }

  private fun handleLoadRewardedAd(call: MethodCall, result: MethodChannel.Result) {
    val adUnitIdentifier = call.argument<String>("adUnitIdentifier")
    val id = adUnitIdentifier?.toIntOrNull()
    if (adUnitIdentifier == null || id == null) {
      result.error("EzoicAds", "Invalid adUnitIdentifier: $adUnitIdentifier", null); return
    }
    if (rewardedAds.containsKey(adUnitIdentifier) || loadingRewarded.contains(adUnitIdentifier)) {
      result.error(
        "EzoicAds", "An ad is already loaded/loading for ad unit $adUnitIdentifier", null); return
    }
    loadingRewarded.add(adUnitIdentifier)
    EzoicRewardedAd.load(appContext, id) { r ->
      loadingRewarded.remove(adUnitIdentifier)
      r.onSuccess { ad ->
        ad.listener = makeListener(adUnitIdentifier)
        rewardedAds[adUnitIdentifier] = ad
        rewardedChannels.getOrPut(adUnitIdentifier) {
          MethodChannel(messenger, "com.ezoic/ezoic_rewarded_ad_$adUnitIdentifier")
        }
        result.success(null)
      }.onFailure { e ->
        result.error("EzoicAds", e.message ?: "Rewarded ad failed to load", e.toString())
      }
    }
  }

  private fun handleShowRewardedAd(call: MethodCall, result: MethodChannel.Result) {
    val adUnitIdentifier = call.argument<String>("adUnitIdentifier")
    val ad = if (adUnitIdentifier != null) rewardedAds[adUnitIdentifier] else null
    if (adUnitIdentifier == null || ad == null) {
      result.error("EzoicAds", "Rewarded ad not loaded for $adUnitIdentifier", null); return
    }
    if (pendingShows.containsKey(adUnitIdentifier)) {
      result.error(
        "EzoicAds", "A show is already in progress for ad unit $adUnitIdentifier", null); return
    }
    val currentActivity = activity
    if (currentActivity == null) {
      result.error("EzoicAds", "No current Activity to present the rewarded ad", null); return
    }

    val show = RewardShow(result)
    pendingShows[adUnitIdentifier] = show
    currentActivity.runOnUiThread {
      ad.show(currentActivity) { reward -> show.reward = reward }
    }
  }

  private fun makeListener(adUnitIdentifier: String) =
    object : EzoicRewardedAdListenerAdapter() {
      override fun onRewardedAdShown(rewardedAd: EzoicRewardedAd) {
        emit(adUnitIdentifier, "onShown")
      }

      override fun onRewardedAdFailedToShow(rewardedAd: EzoicRewardedAd, error: EzoicError) {
        emit(adUnitIdentifier, "onFailedToShow", mapOf("message" to error.message, "code" to error.code))
        cleanup(adUnitIdentifier)
        val pending = pendingShows.remove(adUnitIdentifier)
        if (pending != null && !pending.settled) {
          pending.settled = true
          pending.result.error("EzoicAds", error.message, null)
        }
      }

      override fun onRewardedAdImpression(rewardedAd: EzoicRewardedAd) {
        emit(adUnitIdentifier, "onImpression")
      }

      override fun onRewardedAdClicked(rewardedAd: EzoicRewardedAd) {
        emit(adUnitIdentifier, "onClicked")
      }

      override fun onUserEarnedReward(rewardedAd: EzoicRewardedAd, reward: EzoicReward) {
        emit(adUnitIdentifier, "onUserEarnedReward", mapOf("type" to reward.type, "amount" to reward.amount))
        pendingShows[adUnitIdentifier]?.reward = reward
      }

      override fun onRewardedAdDismissed(rewardedAd: EzoicRewardedAd) {
        emit(adUnitIdentifier, "onDismissed")
        val pending = pendingShows.remove(adUnitIdentifier)
        cleanup(adUnitIdentifier)
        if (pending != null && !pending.settled) {
          pending.settled = true
          val reward = pending.reward
          pending.result.success(
            mapOf(
              "earned" to (reward != null),
              "type" to (reward?.type ?: ""),
              "amount" to (reward?.amount ?: 0)
            )
          )
        }
      }
    }

  private fun cleanup(adUnitIdentifier: String) {
    rewardedAds.remove(adUnitIdentifier)
  }

  private fun emit(adUnitIdentifier: String, method: String, args: Any? = null) {
    val channel = rewardedChannels[adUnitIdentifier] ?: return
    mainHandler.post { channel.invokeMethod(method, args) }
  }

  private fun handleLoadInterstitialAd(call: MethodCall, result: MethodChannel.Result) {
    val adUnitIdentifier = call.argument<String>("adUnitIdentifier")
    val id = adUnitIdentifier?.toIntOrNull()
    if (adUnitIdentifier == null || id == null) {
      result.error("EzoicAds", "Invalid adUnitIdentifier: $adUnitIdentifier", null); return
    }
    if (interstitialAds.containsKey(adUnitIdentifier) || loadingInterstitial.contains(adUnitIdentifier)) {
      result.error(
        "EzoicAds", "An ad is already loaded/loading for ad unit $adUnitIdentifier", null); return
    }
    loadingInterstitial.add(adUnitIdentifier)
    EzoicInterstitialAd.load(appContext, id) { r ->
      loadingInterstitial.remove(adUnitIdentifier)
      r.onSuccess { ad ->
        ad.listener = makeInterstitialListener(adUnitIdentifier)
        interstitialAds[adUnitIdentifier] = ad
        interstitialChannels.getOrPut(adUnitIdentifier) {
          MethodChannel(messenger, "com.ezoic/ezoic_interstitial_ad_$adUnitIdentifier")
        }
        result.success(null)
      }.onFailure { e ->
        result.error("EzoicAds", e.message ?: "Interstitial ad failed to load", e.toString())
      }
    }
  }

  private fun handleShowInterstitialAd(call: MethodCall, result: MethodChannel.Result) {
    val adUnitIdentifier = call.argument<String>("adUnitIdentifier")
    val ad = if (adUnitIdentifier != null) interstitialAds[adUnitIdentifier] else null
    if (adUnitIdentifier == null || ad == null) {
      result.error("EzoicAds", "Interstitial ad not loaded for $adUnitIdentifier", null); return
    }
    if (pendingInterstitialShows.containsKey(adUnitIdentifier)) {
      result.error(
        "EzoicAds", "A show is already in progress for ad unit $adUnitIdentifier", null); return
    }
    val currentActivity = activity
    if (currentActivity == null) {
      result.error("EzoicAds", "No current Activity to present the interstitial ad", null); return
    }

    pendingInterstitialShows[adUnitIdentifier] = InterstitialShow(result)
    // Native show(activity) has no completion lambda, so the show promise is
    // settled from the listener (dismiss = resolve, failed-to-show = reject).
    currentActivity.runOnUiThread {
      ad.show(currentActivity)
    }
  }

  private fun makeInterstitialListener(adUnitIdentifier: String) =
    object : EzoicInterstitialAdListenerAdapter() {
      override fun onInterstitialAdShown(interstitialAd: EzoicInterstitialAd) {
        emitInterstitial(adUnitIdentifier, "onShown")
      }

      override fun onInterstitialAdFailedToShow(interstitialAd: EzoicInterstitialAd, error: EzoicError) {
        emitInterstitial(adUnitIdentifier, "onFailedToShow", mapOf("message" to error.message, "code" to error.code))
        cleanupInterstitial(adUnitIdentifier)
        val pending = pendingInterstitialShows.remove(adUnitIdentifier)
        if (pending != null && !pending.settled) {
          pending.settled = true
          pending.result.error("EzoicAds", error.message, error.code)
        }
      }

      override fun onInterstitialAdImpression(interstitialAd: EzoicInterstitialAd) {
        emitInterstitial(adUnitIdentifier, "onImpression")
      }

      override fun onInterstitialAdClicked(interstitialAd: EzoicInterstitialAd) {
        emitInterstitial(adUnitIdentifier, "onClicked")
      }

      override fun onInterstitialAdDismissed(interstitialAd: EzoicInterstitialAd) {
        emitInterstitial(adUnitIdentifier, "onDismissed")
        val pending = pendingInterstitialShows.remove(adUnitIdentifier)
        cleanupInterstitial(adUnitIdentifier)
        if (pending != null && !pending.settled) {
          pending.settled = true
          pending.result.success(null)
        }
      }
    }

  private fun cleanupInterstitial(adUnitIdentifier: String) {
    interstitialAds.remove(adUnitIdentifier)
  }

  private fun emitInterstitial(adUnitIdentifier: String, method: String, args: Any? = null) {
    val channel = interstitialChannels[adUnitIdentifier] ?: return
    mainHandler.post { channel.invokeMethod(method, args) }
  }

  private fun handleLoadInstreamAd(call: MethodCall, result: MethodChannel.Result) {
    val adUnitIdentifier = call.argument<String>("adUnitIdentifier")
    val id = adUnitIdentifier?.toIntOrNull()
    if (adUnitIdentifier == null || id == null) {
      result.error("EzoicAds", "Invalid adUnitIdentifier: $adUnitIdentifier", null); return
    }
    // Reject overlapping loads for this ad unit: the native load silently
    // no-ops while loading (both a second load on the same instance and a load
    // on a fresh Dart instance sharing the id), which would hang the future.
    if (loadingInstream.containsKey(id)) {
      result.error(
        "EzoicAds", "An instream ad is already loading for ad unit $adUnitIdentifier", null); return
    }
    val contentUrl = call.argument<String>("contentUrl")

    // Create-or-reuse: instream is multi-use, so a repeat load on the same id
    // reuses the existing native controller (preserving its tag state).
    val ad = instreamAds.getOrPut(id) { EzoicInstreamAd(id) }

    val holder = InstreamLoad(result)
    loadingInstream[id] = holder
    ad.load(appContext, contentUrl, object : EzoicInstreamAdListener {
      override fun onAdTagReady(adTagUrl: String) {
        mainHandler.post {
          if (!holder.settled) {
            holder.settled = true
            loadingInstream.remove(id)
            holder.result.success(adTagUrl)
          }
        }
      }

      override fun onAdFailedToLoad(error: EzoicError) {
        mainHandler.post {
          if (!holder.settled) {
            holder.settled = true
            loadingInstream.remove(id)
            holder.result.error("EzoicAds", error.message, error.code)
          }
        }
      }
    })
  }

  private fun handleGetInstreamNextAdTagUrl(call: MethodCall, result: MethodChannel.Result) {
    val id = call.argument<String>("adUnitIdentifier")?.toIntOrNull()
    if (id == null) {
      result.error("EzoicAds", "Invalid adUnitIdentifier", null); return
    }
    result.success(instreamAds[id]?.getNextAdTagUrl())
  }

  private fun handleReportInstreamImpression(call: MethodCall, result: MethodChannel.Result) {
    val id = call.argument<String>("adUnitIdentifier")?.toIntOrNull()
    if (id == null) {
      result.error("EzoicAds", "Invalid adUnitIdentifier", null); return
    }
    // revenueUsd may arrive as an Int over the codec when whole; take it as a
    // Number and coerce to the native Double? the SDK expects.
    val revenueUsd = call.argument<Number>("revenueUsd")?.toDouble()
    instreamAds[id]?.reportImpression(revenueUsd)
    result.success(null)
  }

  private fun handleDestroyInstreamAd(call: MethodCall, result: MethodChannel.Result) {
    val id = call.argument<String>("adUnitIdentifier")?.toIntOrNull()
    if (id == null) {
      result.error("EzoicAds", "Invalid adUnitIdentifier", null); return
    }
    // The native SDKs ignore load callbacks once an ad is destroyed, so a
    // pending load's Flutter result must be settled here or it hangs forever.
    val pending = loadingInstream.remove(id)
    if (pending != null && !pending.settled) {
      pending.settled = true
      pending.result.error("EzoicAds", "Instream ad was destroyed while loading", null)
    }
    instreamAds.remove(id)?.destroy()
    result.success(null)
  }
}
